astropy/utils/state.py

```diff
@@ -7,7 +7,79 @@
 __all__ = ['ScienceState']
 
 
-class ScienceState:
+class StateMixinMeta(type):
+    """Metaclass for state information mixin.
+
+    Creates the ``_state`` variable, if doesn't already exist.
+    The namespace is created by the ``__prepare__`` method,
+    so ``_state`` is defined before even a variable definition.
+
+    items in ``_state`` can be accessed as items of the class.
+
+    Examples
+    --------
+    ::
+
+        class MyClass(metaclass=StateMixinMeta):
+            _state["foo"] = "bar"
+
+        assert MyClass["foo"] == "bar"
+        True
+
+    """
+
+    @classmethod
+    def __prepare__(cls, *args):
+        namespace = dict(_state=dict())
+        return namespace
+
+    def __getitem__(cls, key):
+        """
+        Items in ``_state`` can be accessed as items of the class.
+        """
+        return cls._state[key]
+
+
+class _Context(metaclass=StateMixinMeta):
```

Comment:
I'm confused why this class needs the `StateMixinMeta`.

Reply:
I guess it doesn't currently. I was hoping to do another PR after this to extend the ``_Context`` returned by ``set`` to store the old state and redirect all `getitem` to ``self._parent``. This would allow the ``__enter__`` method to return the context for use in `with` statements. Having `_Context` use the same metaclass is more in line with that PR.

```
MyState._value = 1
with MyState.set(value=2) as context:
    assert context._value == MyState._value == 2
assert MyState._value == 1
with pytest.raises(Exception):
    context._value
```

This currently does not work as the context would have ``_value``=1 in `with`.

Reply:
I do like the idea of being able to do `with ctx as ...:` (as follow-up...)

Reply:
- [x] For followup PR.